File: server/src/UserService.ts
import { Player } from '../../client/src/store/Player';
import { State } from '../../client/src/store/State';

type UserIdPair = {
  id: string;
  name: string;
};

export class UserService {
  registeredUsers: Map<String, Player>;

  constructor() {
    this.registeredUsers = new Map();
  }

  signUp(state: State, name: string, loginId: string): string {
    const id = loginId;
    const user = {
      id,
      name,
      currentBet: 80,
      chipBalance: 5000,
      cashBalance: 100,
    } as Player;
    this.registeredUsers.set(id, user);
    return id;
  }

  login(state: State, id: string): State {
    const user = this.lookup(id);
    if (user && !this.isLoggedIn(state, id)) {
      var activeUsers = [...state.activeUsers, user];
      state = { ...state, activeUsers };
    }
    return state;
  }

  logout(state: State, id: string): State {
    const user = this.lookup(id);
    if (user) {
      var activeUsers = state.activeUsers.filter((u) => u.id !== user.id);
      state = { ...state, activeUsers };
    }
    return state;
  }

  lookup(id: string): Player | null {
    const user = this.registeredUsers.get(id);
    if (user) {
      return user;
    }
    console.log('Could not find user with id=' + id);
    return null;
  }

  getRegistered(): Array<Player> {
    var ret = [];
    let i = this.registeredUsers.values();
    var user: Player;
    while ((user = i.next().value)) {
      ret.push({ ...user });
    }
    return ret;
  }

  isLoggedIn(state: State, loginId: string): boolean {
    return state.activeUsers.find((u) => u.id === loginId) !== undefined;
  }

  modifyPlayer(state: State, player: Player) {
    var activeUsers = state.activeUsers.map((p) => {
      if (p.id === player.id) {
        return player;
      }
      return p;
    });
    return { ...state, activeUsers };
  }
}
export const theUserService = new UserService();

File: django/blackjack/store/user_service.py
from blackjack.state.player import Player
from blackjack.state.state import State

class UserService:
    registeredUsers: dict[str, Player]

    def __init__(self) -> None:
        self.registeredUsers = {}

    def signUp(self, state: State, name: str, loginId: str) -> str :
        id = loginId
        newPlayer = Player()

        newPlayer.id = loginId
        newPlayer.name = name
        newPlayer.currentBet = 80
        newPlayer.chipBalance = 5000
        newPlayer.cashBalance = 100
        self.registeredUsers[id] = newPlayer
        return id
    
    def login(self, state: State, id: str) -> State :
        user = self.lookup(id)
        if user and not self.isLoggedIn(state, id):
            activeUsers = state.activeUsers.copy()
            activeUsers.append(user)
            newState = state.clone()
            newState.activeUsers = activeUsers
            return newState
        return state

    def logout(self, state: State, id: str) -> State :
        user = self.lookup(id)
        if user != None:
            activeUsers = list(filter(lambda u: u.id != user.id, state.activeUsers))
        
            newState = state.clone()
            newState.activeUsers = activeUsers
            return newState
        return state

    def lookup(self, id: str) -> Player | None :
        if id in self.registeredUsers:
            return self.registeredUsers[id]

        print(f'Could not find user with id={id}')
        return None

    def isLoggedIn(self, state: State, loginId: str) -> bool :
        return next(filter(lambda u: u.id == loginId, state.activeUsers), None) != None

    def modifyPlayer(self, state: State, player: Player) -> State :
        activeUsers = list(map(lambda p: player if p.id == player.id else p, state.activeUsers))
        self.registeredUsers[player.id] = player

        newState = state.clone()
        newState.activeUsers = activeUsers
        return newState